src/ivio/detail/zlib_file_reader.h

```diff
@@ -69,6 +69,15 @@ struct zlib_reader {
             }
         }
     }
+
+    /**
+     * Function checks if the given buffer contains the magic header of a gz file
+     */
+    static bool checkHeader(std::span<char const> buffer) {
```

Comment:
Here are some suggestions:

```cpp
#include <algorithm>
#include <span>

bool checkHeader(std::span<char const> buffer)
{
        return buffer.size() >= 2u
            && buffer[0] == '\x1f'
            && buffer[1] == '\x8b';
}

bool checkHeader2(std::span<char const> buffer)
{
    static constexpr std::array<char const, 2> magic_header{'\x1f', '\x8b'};
    return buffer.size() >= 2u && std::ranges::equal(buffer.subspan(0, 2), magic_header);
}
```

`checkHeader2` seems to produce a more succinct assembly. Looks like the compiler collapses the two `char` comparison into one, while in `checkHeader` there are two cmp.

https://godbolt.org/z/bModPnbd8

Reply:
Interesting...That is surprising!

Reply:
```
#include <cstring>

static bool checkHeader(std::span<char const> buffer)
{
    static constexpr std::array<char const, 2> magic_header{'\x1f', '\x8b'};
    static constexpr size_t length{magic_header.size()};
    return buffer.size() >= length && std::memcmp(buffer.data(), magic_header.data(), length) == 0;
}
```

Produces the same assembly, and doesn't need ranges. And it's harder to forget adapting the `2` if this function is copy/pasted for other things.

Reply:
> ```c++
> #include <cstring>
> 
> static bool checkHeader(std::span<char const> buffer)
> {
>     static constexpr std::array<char const, 2> magic_header{'\x1f', '\x8b'};
>     static constexpr size_t length{magic_header.size()};
>     return buffer.size() >= length && std::memcmp(buffer.data(), magic_header.data(), length) == 0;
> }
> ```
> 
> Produces the same assembly, and doesn't need ranges. And it's harder to forget adapting the `2` if this function is copy/pasted for other things.

Yes, the '2' was annoying. What do you thing about using std::string_view?
Advantage is, we could use `starts_with` which makes the intend very clear:
```
bool checkHeader6(std::span<char const> buffer)
{
    static constexpr std::string_view magic_header{"\x1f\x8b"};
    return std::string_view{buffer}.starts_with(magic_header);
}
```

Reply:
Or even prettier:
```c++
bool checkHeader7(std::span<char const> buffer)
{
    auto view = std::string_view{buffer.data(), buffer.size()};
    return view.starts_with("\x1f\x8b");
}
```

Reply:
In c++23 we could bring it down to:
```c++
bool checkHeader8(std::span<char const> buffer)
{
    return std::string_view{buffer}.starts_with("\x1f\x8b");
}
```

Reply:
> Or even prettier:
> ```c++
> bool checkHeader7(std::span<char const> buffer)
> {
>     auto view = std::string_view{buffer.data(), buffer.size()};
>     return view.starts_with("\x1f\x8b");
> }
> ```

```cpp
bool checkHeader9(std::string_view /* const? */ buffer)
{
    return buffer.starts_with("\x1f\x8b");
}
```

Should also work. Both span and string_view have a `char const *, size_t` constructor.
The big question is whether all compilers implement the starts_with :D

Reply:
True! I wasn't sure about using std::string_view for stuff that are technically not strings. Sadly it has all the cool functions and std::span not :-(